### ISSUE TITLE
Considered this as a native plugin feature in Webpack

### ISSUE THREAD

**Issue by TheLarkInn** [amireh/happypack#52]
Is there a reason you haven't submitted this parallel file loading as a feature or plugin into webpack? @sokra @jhnns and I were talking about way to have faster webpack builds and this has the potential with a bit of cleanup. Have you considered submitting this plugin for a PR. We are also looking into a strategy that caches the dependency AST as loaders run and persisting them for chained loaders tied to it etc.

**Comment by jhnns** [comment]
To be honest, I don't know where the performance improvements of happypack come from. Webpack already invokes loaders in parallel. And when a loader sets the `cacheable()` flag (which all popular loaders do), webpack cache's the loader's output as well.

However, I really appreciate this effort and I think every performance improvement should be merged into core.

**Comment by amireh** [comment]
I did mention it around the time HappyPack was made public in https://github.com/webpack/webpack/issues/1810 but it never really garnered much interest so I didn't push it.

What would the benefits be if we choose to push this to core and not use a plugin?

My concern with trying to add this to core is that, with webpack's current design, we will never have full support for all loader APIs since not everything can be shared cross processes. In particular, I'm talking about hooking into compilation and compiler instance states from loaders. Since this is currently a plugin, it's ok for us to say "well, this won't work if your loaders use X or Y" but it's a different story if webpack itself tries to provide this functionality - then it is bereft of that freedom.

@jhnns IIRC webpack uses `async` to apply loaders in parallel but that's still bound to the same node process (and the same event loop) which isn't the same as launching multiple _processes_ which is what HappyPack does. Does that make sense?

**Comment by jhnns** [comment]
> What would the benefits be if we choose to push this to core and not use a plugin?

Everyone should benefit from performance improvements 😁 

> My concern with trying to add this to core is that, with webpack's current design, we will never have full support for all loader APIs since not everything can be shared cross processes. In particular, I'm talking about hooking into compilation and compiler instance states from loaders. Since this is currently a plugin, it's ok for us to say "well, this won't work if your loaders use X or Y" but it's a different story if webpack itself tries to provide this functionality - then it is bereft of that freedom.

That might be a problem. However, a loader has no official access to the compilation by design and I don't know any popular loader that does this. So this should be no problem. We could also add a getter to the `_compilation` and `_compiler` property to warn the user that a loader is trying to access them.

> webpack uses async to apply loaders in parallel but that's still bound to the same node process (and the same event loop) which isn't the same as launching multiple processes which is what HappyPack does. Does that make sense?

Not really. Node's concept is based on the assumption that a single process with an event loop handles IO more efficiently than multiple processes. As long as you're not doing synchronous stuff (which means that the process is waiting for a single result without doing anything else) I would expect a single process to be at least as fast as multiple processes.

> Edit: also, re the cacheable() flag, that is still bound to the process, it won't persist on subsequent runs (which is OK since it's meant to make incremental builds faster, but happypack caches to disk which takes things one step further.)

This increases the initial build time indeed. But what if you've updated a loader or plugin in the meantime? Do you track dependencies as well? [We already had a long discussion about this](https://github.com/webpack/webpack/issues/250) and we haven't found a good way yet.

I'll try HappyPack in my projects to check if there is a performance improvement and maybe to find out, where the improvement is coming from. Thanks for your hard work! 👍

**Comment by amireh** [comment]
> That might be a problem. However, a loader has no official access to the compilation by design and I don't know any popular loader that does this. So this should be no problem. We could also add a getter to the _compilation and _compiler property to warn the user that a loader is trying to access them.

Hacky access still gave people access so they use it :grin: - the typescript loaders all seem to rely on `compilation` and compiler hooks which is making it hard to support them in this architecture.

But all in all, I think webpack would benefit from the modularity if we can isolate the application of loaders from the core process (like what's happening here) if for anything but the separation of concerns. I also wish that we converge on a single loader application implementation, because right now happypack tries to re-do what webpack already does (see https://github.com/amireh/happypack/blob/master/lib/applyLoaders.js) in terms of how it applies the loaders. (If I'm not mistaken, I saw that sokra was making a library for this which may end up being used in webpack but it was still tied to webpack internals and made some assumptions that makes it inapplicable on a distributed architecture..)

> Not really. Node's concept is based on the assumption that a single process with an event loop handles IO more efficiently than multiple processes. As long as you're not doing synchronous stuff (which means that the process is waiting for a single result without doing anything else) I would expect a single process to be at least as fast as multiple processes.

I believe that's really it. In our own use case, babel transformations were taking most of the time and those (presumably) are synchronous. This seems to also be the case for any loader that performs some form of AST analysis which tend to be slow (cpu and not IO-bound).

> But what if you've updated a loader or plugin in the meantime? Do you track dependencies as well?

I didn't try to solve that, instead, I just made it easy to nuke the entire cache when you know it's gone bad (just do an `rm -r .happypack`) and then you can also specify some variables that constitute a cache "context", which if changes between builds, the cache will considered to be invalid (useful for environment variables and such). It is silly and perhaps a little irresponsible, I know! But it works for most use cases and reaps some nice benefits when it _does_ work so :smile: 

It's funny reading through that thread you linked, because at some point in the past I had my own take at making a serializable cache for webpack, which did work to some extent, but then I found the (de)hydration costs to be too high to make it feasible. I was probably caching some unnecessary information. However, it's cool to see that others have been working on this, I hope it is solved one day!

Thanks for the work on webpack.. it's one beast of a software! xD

**Comment by jhnns** [comment]
> Hacky access still gave people access so they use it 😁 - the typescript loaders all seem to rely on compilation and compiler hooks which is making it hard to support them in this architecture.

Yeah, I wish sokra had not added it... . Didn't know about the typescript-loader, though.

> But all in all, I think webpack would benefit from the modularity if we can isolate the application of loaders from the core process (like what's happening here) if for anything but the separation of concerns

Definitely.

> I believe that's really it. In our own use case, babel transformations were taking most of the time and those (presumably) are synchronous. This seems to also be the case for any loader that performs some form of AST analysis which tend to be slow (cpu and not IO-bound).

Yeah, I think multi-core CPUs are important here because then you can lift the heavy JS work on multiple cores. Maybe that's the cause. In this case, however, you cannot benefit from cached FS access. But I don't know, maybe the OS FS cache is sufficient here.

> I didn't try to solve that, instead, I just made it easy to nuke the entire cache when you know it's gone bad

That's ok, but I think it should be opt-in. When you don't know about caching, this can bite you.

**Comment by amireh** [comment]
> That's ok, but I think it should be opt-in. When you don't know about caching, this can bite you.

It is on by default, although you can totally disable the cache if you want. We can probably make it a bit more sophisticated to catch some common scenarios (like updating node modules) or just have the documentation state this and leave it up to the user like it is now. In practice, the caching bit is not the main gain from happypack, and it would still be a big improvement to use it with caching disabled altogether.

Anyway, let us know when you try it out and how you think we should proceed. I'd be happy to help.

**Comment by csvan** [comment]
+1, would love to see this as a native component in webpack. It's almost mandatory for large project, and much easier to setup than DLLs.

**Comment by TheLarkInn** [comment]
I think finding a creative API for execution is a strong interest to me.

**Comment by amireh** [comment]
Trying to tidy things up around here so I'll be closing this, thanks all! :) (FWIW caching has been removed as of v4, this is now only concerned with parallelism)
